### PR TITLE
Add an inference example for Tacotron2

### PR DESCRIPTION
**Pull request by yangarbiter**
An example for doing Tacotron2 inferencing.
#1642 #1648 should be merged before this can work.

### CODE REVIEW

File: examples/pipeline_tacotron2/README.md [hunk 1]
```diff
@@ -141,3 +141,95 @@ python train.py \
     --checkpoint-path ./english_phonemes_wavernn_ckpt.pth \
     --dataset-path ./
 ```
+
+
+## Text-to-speech pipeline
+
+Here we present an example of how to use Tacotron2 to generate audio from text.
+For the text-to-speech pipeline, with an input text, 
+Tacotron2 will first generate the corresponding mel spectrogram.
```
Comment (mthrok) [inline examples/pipeline_tacotron2/README.md:150]: nit: for those who do not know how TTS works, giving a brief description of process (text preprocessing, feature extraction, spectrogram generation, time-domain conversion) would be nice. (if not done so)
Maybe it is better suited for tutorial.

File: examples/pipeline_tacotron2/inference.py [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+import argparse
+import os
+import random
+import sys
+
+import torch
+import torchaudio
+import numpy as np
+from torchaudio.prototype.tacotron2 import Tacotron2
+from torchaudio.models.wavernn import _MODEL_CONFIG_AND_URLS
```
Comment (mthrok) [inline examples/pipeline_tacotron2/inference.py:11]: Can you put this into `_parse_args` so that it will be (somewhat) hidden?

File: examples/pipeline_tacotron2/inference.py [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+        if args.jit:
+            vocoder = torch.jit.script(vocoder)
+
+        waveform = vocoder(mel_specgram.cpu())
+
+    torchaudio.save(args.output_path, waveform, args.sample_rate)
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description='TTS Generator')
```
Comment (mthrok) [inline examples/pipeline_tacotron2/inference.py:290]: What I often do when writing a script is;
1. Add a module-level docstring.
2. Refer the docstring `__doc__` from `ArgumentParser(description=__doc__)`.
That way, one can write detailed docstring which is accessible from CLI as well.

(Also I would put the construction of parer in `parse_args` as well.)

File: examples/pipeline_tacotron2/inference.py [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+
+def main(args):
+    torch.manual_seed(0)
+    random.seed(0)
+    np.random.seed(0)
+
+    device = "cuda" if torch.cuda.is_available() else "cpu"
+
+    n_symbols = len(get_symbol_list(args.text_preprocessor))
+    text_preprocessor = partial(
```
Comment (mthrok) [inline examples/pipeline_tacotron2/inference.py:186]: Similar to https://github.com/pytorch/audio/pull/1668/files#r686174534, turning this into a class will make the code more maintainable.

File: examples/pipeline_tacotron2/inference.py [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+        "--wavernn-batch-overlap",
+        default=550,
+        type=int,
+        help="The overlapping time steps between batches. Only used when batch inference is used",
+    )
+
+    return parser
+
+
+def unwrap_distributed(state_dict):
```
Comment (mthrok) [inline examples/pipeline_tacotron2/inference.py:164]: I feel we should update the pre-trained model, so that the structure is unwrapped, if that's not BC-breaking. Thoughts?

Comment (yangarbiter) [reply examples/pipeline_tacotron2/inference.py:164]: For this part, it is used to load the checkpoints after the user train a model with `train.py`. So I think we can keep it this way (assuming the user trained the model from scratch themselfves.

To load our pertained model build in the library, I'll create a separate argument for it.

File: examples/pipeline_tacotron2/inference.py (hunk 4 above)
Comment (yangarbiter) [reply examples/pipeline_tacotron2/inference.py:205]: I'll submit a separate PR for this. I've created an issue #1710 to keep track of this.

File: examples/pipeline_tacotron2/inference.py (hunk 3 above)
Comment (yangarbiter) [reply examples/pipeline_tacotron2/inference.py]: Thanks for the reminder, I've move the parser into `parse_args` and wrote a simple module docstring.

File: examples/pipeline_tacotron2/inference.py (hunk 2 above)
Comment (yangarbiter) [reply examples/pipeline_tacotron2/inference.py]: Good suggestion! I've moved it into `parse_args` [here](https://github.com/pytorch/audio/pull/1654/files#diff-190b61efc2a83079d005a5da48a9fde8b28a937a8a653a8b4ede3b792f742384R31)

File: examples/pipeline_tacotron2/README.md (hunk 1 above)
Comment (yangarbiter) [reply examples/pipeline_tacotron2/README.md]: I've updated the README [here](https://github.com/pytorch/audio/pull/1654/files#diff-a17fd0fdb11c8336af63efc614215c0ef98e67d810b77c5c5e8d87a012ca64f1R148-R157) and hopefully its a bit clearer.

And yes, when the tutorial is ready we should definitely add the link here.

File: examples/pipeline_tacotron2/inference.py [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+    if args.jit:
+        tacotron2 = torch.jit.script(tacotron2)
+
+    sequences, lengths = prepare_input_sequence([args.input_text],
+                                                text_processor=text_preprocessor)
+    sequences, lengths = sequences.long().to(device), lengths.long().to(device)
+    with torch.no_grad():
+        mel_specgram, _, _ = tacotron2.infer(sequences, lengths)
+
+    if args.vocoder == "nvidia_waveglow":
```
Comment (mthrok) [inline examples/pipeline_tacotron2/inference.py:235]: This `if ~ elif ~ else` clause is huge, which makes it difficult to follow the flow. Extracting each logic into helper functions help.

File: examples/pipeline_tacotron2/inference.py [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+
+        if args.jit:
+            raise ValueError("Vocoder option `nvidia_waveglow is not jittable.")
+
+        with torch.no_grad():
+            waveform = waveglow.infer(mel_specgram).cpu()
+
+    elif args.vocoder == "wavernn":
+        from torchaudio.models import wavernn
+        sys.path.append(os.path.join(os.path.dirname(__file__), "../pipeline_wavernn"))
```
Comment (mthrok) [inline examples/pipeline_tacotron2/inference.py:249]: Not in the scope of this PR, but I think putting components for wavernn and tacotoron2 under the same directory, say `examples/tts` would allow to organize the code better.

**Review by mthrok — APPROVED** [review]

File: examples/pipeline_tacotron2/inference.py (hunk 7 above)
Comment (yangarbiter) [reply examples/pipeline_tacotron2/inference.py]: I've created [an issue](https://github.com/pytorch/audio/issues/1712) to keep track of this.

File: examples/pipeline_tacotron2/inference.py (hunk 5 above)
Comment (yangarbiter) [reply examples/pipeline_tacotron2/inference.py:179]: I've created an issue [here](https://github.com/pytorch/audio/issues/1713) for future improvement.
